File: python-algo/leetcode/problem_345.py
"""345. Reverse Vowels of a String
https://leetcode.com/problems/reverse-vowels-of-a-string/
"""


class Solution:
    def reverse_vowels(self, s: str) -> str:
        vowels = ('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U')
        ans = list(s)
        i, j = 0, len(s) - 1
        while i < j:
            if ans[i] not in vowels:
                i += 1
            if ans[j] not in vowels:
                j -= 1
            if ans[i] in vowels and ans[j] in vowels:
                ans[i], ans[j] = ans[j], ans[i]
                i += 1
                j -= 1
        return ''.join(ans)

File: java-algo/src/main/java/com/leetcode/Problem345.java
package com.leetcode;

/**
 * 345. Reverse Vowels of a String
 * https://leetcode.com/problems/reverse-vowels-of-a-string/
 */
public class Problem345 {
    public String reverseVowels(String s) {
        char[] cs = s.toCharArray();
        int i = 0, j = cs.length - 1;
        while (i < j) {
            while (i < j && isVowel(cs[i])) {
                i++;
            }
            while (i < j && isVowel(cs[j])) {
                j--;
            }
            if (i < j) {
                char tmp = cs[i];
                cs[i] = cs[j];
                cs[j] = tmp;
                i++;
                j--;
            }
        }
        return new String(cs);
    }

    private boolean isVowel(char c) {
        return c != 'a' && c != 'e' && c != 'i' && c != 'o' && c != 'u' && c != 'A' && c != 'E' && c != 'I' && c != 'O' && c != 'U';
    }
}
